lib/internal/util/comparisons.js

```diff
@@ -147,84 +202,114 @@ function looseDeepEqual(val1, val2, memos) {
-    if (val1.message !== val2.message || val1.name !== val2.name)
-      return false;
-  }
   const val1Tag = objectToString(val1);
   const val2Tag = objectToString(val2);
   if (val1Tag === val2Tag) {
-    if (!isObjectOrArrayTag(val1Tag) && isArrayBufferView(val1)) {
-      return areSimilarTypedArrays(val1, val2,
-                                   isFloatTypedArrayTag(val1Tag) ?
-                                     Infinity : 300);
+    if (isObjectOrArrayTag(val1Tag)) {
+      return keyCheck(val1, val2, kLoose, memos, kNoIterator);
+    }
+    if (isArrayBufferView(val1)) {
+      if (isFloatTypedArrayTag(val1Tag)) {
+        return areSimilarFloatArrays(val1, val2);
+      }
+      return areSimilarTypedArrays(val1, val2);
+    }
+    if (isDate(val1) && isDate(val2)) {
+      return getTime(val1) === getTime(val2);
+    }
+    if (isRegExp(val1) && isRegExp(val2)) {
+      return areSimilarRegExps(val1, val2);
+    }
+    if (val1 instanceof Error && val2 instanceof Error) {
+      if (val1.message !== val2.message || val1.name !== val2.name)
+        return false;
     }
   // Ensure reflexivity of deepEqual with `arguments` objects.
   // See https://github.com/nodejs/node-v0.x-archive/pull/7178
   } else if (isArguments(val1Tag) || isArguments(val2Tag)) {
     return false;
   }
-  return keyCheck(val1, val2, false, memos);
+  if (isSet(val1)) {
+    if (!isSet(val2) || val1.size !== val2.size) {
+      return false;
+    }
+    return keyCheck(val1, val2, kLoose, memos, kIsSet);
+  } else if (isMap(val1)) {
+    if (!isMap(val2) || val1.size !== val2.size) {
+      return false;
+    }
+    return keyCheck(val1, val2, kLoose, memos, kIsMap);
+  } else if (isSet(val2) || isMap(val2)) {
+    return false;
+  }
+  return keyCheck(val1, val2, kLoose, memos, kNoIterator);
 }
 
-function keyCheck(val1, val2, strict, memos, lengthA, lengthB) {
+function getEnumerables(val, keys) {
+  return keys.filter((k) => propertyIsEnumerable(val, k));
+}
+
+function keyCheck(val1, val2, strict, memos, iterator, aKeys) {
   // For all remaining Object pairs, including Array, objects and Maps,
   // equivalence is determined by having:
   // a) The same number of owned enumerable properties
   // b) The same set of keys/indexes (although not necessarily the same order)
   // c) Equivalent values for every corresponding key/index
   // d) For Sets and Maps, equal contents
   // Note: this accounts for both named and indexed properties on Arrays.
-  var aKeys = Object.keys(val1);
-  var bKeys = Object.keys(val2);
-  var i;
+  if (arguments.length === 5) {
+    aKeys = objectKeys(val1);
+    const bKeys = objectKeys(val2);
 
-  // The pair must have the same number of owned properties.
-  if (aKeys.length !== bKeys.length)
-    return false;
+    // The pair must have the same number of owned properties.
+    if (aKeys.length !== bKeys.length) {
+      return false;
+    }
+  }
+
+  // Cheap key test
+  let i = 0;
```

Comment:
Where is `i` used outside of the loop here?  It confused me since I expected it to be used at some capacity elsewhere. I don't think making `i` local to the loop would hurt performance.

Reply:
It is used in two loops. I could define a second variable but this seemed nicer.

Reply:
I find it a bit more confusing than defining it in the loop for the reason above but don't feel strongly about it - so if you think it's nicer I concede.